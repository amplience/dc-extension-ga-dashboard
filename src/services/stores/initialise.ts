import {
  contentItemFilter,
  editionFilter,
  slotFilter,
} from '../../stores/ga-query-filters';
import {
  breakdown,
  contentItemIdMapping,
  editionIdMapping,
  gaApiKey,
  gaClientEmail,
  gaClientId,
  gaViewId,
  slotIdMapping,
} from '../../stores/google-analytics';
import { currencyCode, locale } from '../../stores/localization';
import type { ExtensionInstallation } from '../extension-sdk/extension-sdk.service';

export function initialiseStores(config: ExtensionInstallation): void {
  gaViewId.set(config?.googleAnalyticsViewId || '');
  gaClientId.set(config?.googleAnalyticsClientId || '');
  gaClientEmail.set(config?.googleAnalyticsClientEmail);
  gaApiKey.set(config?.googleAnalyticsKey || '');

  contentItemIdMapping.set(config?.mappings?.contentItemId);
  editionIdMapping.set(config?.mappings?.editionId);
  slotIdMapping.set(config?.mappings?.slotId);

  contentItemFilter.set(config?.filters?.contentItemFilter);
  editionFilter.set(config?.filters?.editionFilter);
  slotFilter.set(config?.filters?.slotFilter);

  breakdown.set(
    config?.breakdown || {
      dimension: 'ga:deviceCategory',
      title: 'Device Breakdown',
    }
  );

  if (config?.localization?.locale) {
    locale.set(config?.localization?.locale);
  }

  if (config?.localization?.currencyCode) {
    currencyCode.set(config?.localization?.currencyCode);
  }
}
