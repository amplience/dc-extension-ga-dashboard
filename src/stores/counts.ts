import { writable, derived } from 'svelte/store';
import type {
  SearchIndexUsersCount,
  SearchIndexSearchesCount,
  SearchIndexNoResultsRate,
} from 'dc-management-sdk-js';

export interface Counts {
  usersCount?: SearchIndexUsersCount;
  searchesCount?: SearchIndexSearchesCount;
  noResultsRate?: SearchIndexNoResultsRate;
}

export const counts = writable<Counts>({});

export const searchesPerUser = derived(counts, ($counts) => {
  return $counts?.usersCount?.count && $counts?.searchesCount?.count
    ? Math.round(
        ($counts?.searchesCount?.count / $counts?.usersCount?.count) * 1e1
      ) / 1e1
    : 0;
});
