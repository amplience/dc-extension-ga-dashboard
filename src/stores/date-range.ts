import { formatDateAsISOString } from '../utils/date-format';
import { persistedWritable } from './persisted-writable';

export interface DateRange {
  from: string;
  to: string;
}

export const DAY = 24 * 60 * 60 * 1000;
export const NOW = new Date().setHours(0, 0, 0, 0).valueOf();

export const INITIAL_DATE_RANGE = {
  from: formatDateAsISOString(new Date(NOW - DAY * 7)),
  to: formatDateAsISOString(new Date(NOW - DAY)),
};

export const dateRange = persistedWritable<DateRange>(
  'date-range',
  INITIAL_DATE_RANGE
);
