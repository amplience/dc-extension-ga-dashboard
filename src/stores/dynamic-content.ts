import type { DynamicContent, Hub } from 'dc-management-sdk-js';
import { writable } from 'svelte/store';

export const client = writable<DynamicContent>(null);

export const hub = writable<Hub>(null);
