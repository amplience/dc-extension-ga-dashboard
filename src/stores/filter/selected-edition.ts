import type { Edition } from 'dc-management-sdk-js';
import { persistedWritable } from '../persisted-writable';

export const selectedEdition = persistedWritable<Edition>(
  'selectedEdition',
  null
);
