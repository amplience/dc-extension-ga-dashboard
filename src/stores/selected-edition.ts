import { writable as persistedWritable } from './persisted-store';
import type { Edition } from 'dc-management-sdk-js';

export const selectedEdition = persistedWritable<Edition>(
  'selectedEdition',
  null
);
