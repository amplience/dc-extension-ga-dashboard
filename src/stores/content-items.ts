import { get, writable } from 'svelte/store';
import { client } from './dynamic-content';

function createContentItems() {
  const { subscribe, set, update } = writable([]);

  return {
    subscribe,
    fetch: async (id) => {
      const cachedContentItem = get(contentItems).find(
        (item) => item.id === id
      );
      if (cachedContentItem) {
        return cachedContentItem;
      }
      const contentItem = await get(client).contentItems.get(id);
      update((n) => [...n, ...[contentItem]]);
      return contentItem;
    },
    reset: () => set([]),
  };
}

export const contentItems = createContentItems();
