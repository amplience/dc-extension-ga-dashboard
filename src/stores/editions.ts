import { get, writable } from 'svelte/store';
import { client } from './dynamic-content';

function createEditions() {
  const { subscribe, set, update } = writable([]);

  return {
    subscribe,
    fetch: async (id) => {
      const cachedEdition = get(editions).find((item) => item.id === id);
      if (cachedEdition) {
        return cachedEdition;
      }
      const edition = await get(client).editions.get(id);
      update((n) => [...n, ...[edition]]);
      return edition;
    },
    reset: () => set([]),
  };
}

export const editions = createEditions();
