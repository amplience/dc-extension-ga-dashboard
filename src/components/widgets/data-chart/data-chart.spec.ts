import { render } from '@testing-library/svelte';
import DataChart from './data-chart.svelte';
import { tick } from 'svelte';
import { dateRange } from '../../../stores/date-range';
import {
  ChartType,
  insertDataChart,
  RequestTimeout,
} from '../../../stores/gapi';
import {
  editionIdMapping,
  setGaConfig,
} from '../../../stores/google-analytics';
import { selectedEdition } from '../../../stores/selected-edition';
import { Edition } from 'dc-management-sdk-js';

jest.mock('../../../stores/gapi');

describe('DataChart', () => {
  beforeEach(() => {
    selectedEdition.set(null);
    dateRange.set({ from: '2020-11-01', to: '2020-11-02' });
    setGaConfig({
      googleAnalyticsViewId: '1234567890',
      googleAnalyticsClientId: '1234567890',
      mappings: {
        contentItemId: 'dimension1',
        editionId: 'dimension2',
        slotId: 'dimension3',
      },
    });
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  describe('overview chart', () => {
    const dataChartOptions = {
      className: 'overview',
      title: 'Overview',
      dimensions: 'ga:date',
      chartType: ChartType.LINE,
    };
    it('should render the DataChart component', async () => {
      const { container } = render(DataChart, dataChartOptions);
      await tick();

      expect(insertDataChart).toHaveBeenCalledTimes(1);
      expect((insertDataChart as jest.Mock).mock.calls).toMatchSnapshot();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should update the chart when the dateRange changes', async () => {
      render(DataChart, dataChartOptions);
      await tick();

      dateRange.set({ from: '2020-01-01', to: '2020-01-02' });
      await tick();

      expect(insertDataChart).toHaveBeenCalledTimes(2);
      expect((insertDataChart as jest.Mock).mock.calls).toMatchSnapshot();
    });

    it('should update the chart when the selectedEdition changes', async () => {
      editionIdMapping.set('dimension2');
      render(DataChart, dataChartOptions);
      await tick();

      selectedEdition.set(new Edition({ id: 'editionId' }));
      await tick();

      expect(insertDataChart).toHaveBeenCalledTimes(2);
      expect((insertDataChart as jest.Mock).mock.calls).toMatchSnapshot();
    });

    it('should retry when we get a timeout error', async () => {
      editionIdMapping.set('dimension2');
      (insertDataChart as jest.Mock).mockImplementation(() =>
        Promise.reject(new RequestTimeout('GAPI request timeout'))
      );
      render(DataChart, dataChartOptions);
      await tick();
      await tick();
      (insertDataChart as jest.Mock).mockImplementation(() =>
        Promise.resolve([])
      );
      expect(insertDataChart).toHaveBeenCalledTimes(2);
      expect((insertDataChart as jest.Mock).mock.calls).toMatchSnapshot();
    });
  });

  describe('breakdown chart', () => {
    const dataChartOptions = {
      className: 'breakdown',
      title: 'Breakdown',
      dimensions: 'ga:deviceCategory',
      chartType: ChartType.BAR,
    };
    it('should render the DataChart component', async () => {
      const { container } = render(DataChart, dataChartOptions);
      await tick();
      expect(insertDataChart).toHaveBeenCalledTimes(1);
      expect((insertDataChart as jest.Mock).mock.calls).toMatchSnapshot();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should update the chart when the dateRange changes', async () => {
      render(DataChart, dataChartOptions);
      await tick();

      dateRange.set({ from: '2020-01-01', to: '2020-01-02' });
      await tick();

      expect(insertDataChart).toHaveBeenCalledTimes(2);
      expect((insertDataChart as jest.Mock).mock.calls).toMatchSnapshot();
    });

    it('should update the chart when the selectedEdition changes', async () => {
      editionIdMapping.set('dimension2');
      render(DataChart, dataChartOptions);
      await tick();

      selectedEdition.set(new Edition({ id: 'editionId' }));
      await tick();

      expect(insertDataChart).toHaveBeenCalledTimes(2);
      expect((insertDataChart as jest.Mock).mock.calls).toMatchSnapshot();
    });

    it('should retry when we get a timeout error', async () => {
      editionIdMapping.set('dimension2');
      (insertDataChart as jest.Mock).mockImplementation(() =>
        Promise.reject(new RequestTimeout('GAPI request timeout'))
      );
      render(DataChart, dataChartOptions);
      await tick();
      await tick();
      (insertDataChart as jest.Mock).mockImplementation(() =>
        Promise.resolve([])
      );
      expect(insertDataChart).toHaveBeenCalledTimes(2);
      expect((insertDataChart as jest.Mock).mock.calls).toMatchSnapshot();
    });
  });
});
