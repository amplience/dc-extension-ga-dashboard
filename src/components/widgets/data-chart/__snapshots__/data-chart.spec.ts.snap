// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataChart breakdown chart should render the DataChart component 1`] = `
Array [
  Array [
    Object {
      "dimensions": "ga:deviceCategory",
      "end-date": "2020-11-02",
      "filters": null,
      "ids": "ga:1234567890",
      "metrics": "ga:totalEvents,ga:uniqueEvents",
      "start-date": "2020-11-01",
    },
    "ga-chart-breakdown",
    "BAR",
    Object {},
  ],
]
`;

exports[`DataChart breakdown chart should render the DataChart component 2`] = `
<div>
  <section
    class="breakdown svelte-12xsrpj"
  >
    <article
      class="svelte-pntwh0"
    >
       
      <section
        class="widget-header svelte-1kld2ef"
      >
        <div
          class="title svelte-1kld2ef"
        >
          <h2
            class="svelte-1kld2ef"
          >
            Breakdown
          </h2>
           
           
        </div>
         
        <div
          class="actions svelte-1kld2ef"
        />
      </section>
       
      <section
        class="widget-body svelte-4p1s5s"
      >
        <section
          class="ga-chart svelte-12xsrpj"
        >
           
          <div
            id="ga-chart-breakdown"
          />
        </section>
      </section>
    </article>
  </section>
</div>
`;

exports[`DataChart breakdown chart should update the chart when the dateRange changes 1`] = `
Array [
  Array [
    Object {
      "dimensions": "ga:deviceCategory",
      "end-date": "2020-11-02",
      "filters": null,
      "ids": "ga:1234567890",
      "metrics": "ga:totalEvents,ga:uniqueEvents",
      "start-date": "2020-11-01",
    },
    "ga-chart-breakdown",
    "BAR",
    Object {},
  ],
  Array [
    Object {
      "dimensions": "ga:deviceCategory",
      "end-date": "2020-01-02",
      "filters": null,
      "ids": "ga:1234567890",
      "metrics": "ga:totalEvents,ga:uniqueEvents",
      "start-date": "2020-01-01",
    },
    "ga-chart-breakdown",
    "BAR",
    Object {},
  ],
]
`;

exports[`DataChart breakdown chart should update the chart when the selectedEdition changes 1`] = `
Array [
  Array [
    Object {
      "dimensions": "ga:deviceCategory",
      "end-date": "2020-11-02",
      "filters": null,
      "ids": "ga:1234567890",
      "metrics": "ga:totalEvents,ga:uniqueEvents",
      "start-date": "2020-11-01",
    },
    "ga-chart-breakdown",
    "BAR",
    Object {},
  ],
  Array [
    Object {
      "dimensions": "ga:deviceCategory",
      "end-date": "2020-11-02",
      "filters": "ga:dimension2==editionId",
      "ids": "ga:1234567890",
      "metrics": "ga:totalEvents,ga:uniqueEvents",
      "start-date": "2020-11-01",
    },
    "ga-chart-breakdown",
    "BAR",
    Object {},
  ],
]
`;

exports[`DataChart overview chart should render the DataChart component 1`] = `
Array [
  Array [
    Object {
      "dimensions": "ga:date",
      "end-date": "2020-11-02",
      "filters": null,
      "ids": "ga:1234567890",
      "metrics": "ga:totalEvents,ga:uniqueEvents",
      "start-date": "2020-11-01",
    },
    "ga-chart-overview",
    "LINE",
    Object {
      "hAxis": Object {
        "gridlines": Object {
          "units": Object {
            "days": Object {
              "format": Array [
                "dd MMM",
              ],
            },
          },
        },
        "minorGridlines": Object {
          "units": Object {
            "hours": Object {
              "format": Array [
                "",
              ],
            },
            "minutes": Object {
              "format": Array [
                "",
              ],
            },
          },
        },
      },
    },
  ],
]
`;

exports[`DataChart overview chart should render the DataChart component 2`] = `
<div>
  <section
    class="overview svelte-12xsrpj"
  >
    <article
      class="svelte-pntwh0"
    >
       
      <section
        class="widget-header svelte-1kld2ef"
      >
        <div
          class="title svelte-1kld2ef"
        >
          <h2
            class="svelte-1kld2ef"
          >
            Overview
          </h2>
           
           
        </div>
         
        <div
          class="actions svelte-1kld2ef"
        />
      </section>
       
      <section
        class="widget-body svelte-4p1s5s"
      >
        <section
          class="ga-chart svelte-12xsrpj"
        >
           
          <div
            id="ga-chart-overview"
          />
        </section>
      </section>
    </article>
  </section>
</div>
`;

exports[`DataChart overview chart should update the chart when the dateRange changes 1`] = `
Array [
  Array [
    Object {
      "dimensions": "ga:date",
      "end-date": "2020-11-02",
      "filters": null,
      "ids": "ga:1234567890",
      "metrics": "ga:totalEvents,ga:uniqueEvents",
      "start-date": "2020-11-01",
    },
    "ga-chart-overview",
    "LINE",
    Object {
      "hAxis": Object {
        "gridlines": Object {
          "units": Object {
            "days": Object {
              "format": Array [
                "dd MMM",
              ],
            },
          },
        },
        "minorGridlines": Object {
          "units": Object {
            "hours": Object {
              "format": Array [
                "",
              ],
            },
            "minutes": Object {
              "format": Array [
                "",
              ],
            },
          },
        },
      },
    },
  ],
  Array [
    Object {
      "dimensions": "ga:date",
      "end-date": "2020-01-02",
      "filters": null,
      "ids": "ga:1234567890",
      "metrics": "ga:totalEvents,ga:uniqueEvents",
      "start-date": "2020-01-01",
    },
    "ga-chart-overview",
    "LINE",
    Object {
      "hAxis": Object {
        "gridlines": Object {
          "units": Object {
            "days": Object {
              "format": Array [
                "dd MMM",
              ],
            },
          },
        },
        "minorGridlines": Object {
          "units": Object {
            "hours": Object {
              "format": Array [
                "",
              ],
            },
            "minutes": Object {
              "format": Array [
                "",
              ],
            },
          },
        },
      },
    },
  ],
]
`;

exports[`DataChart overview chart should update the chart when the selectedEdition changes 1`] = `
Array [
  Array [
    Object {
      "dimensions": "ga:date",
      "end-date": "2020-11-02",
      "filters": null,
      "ids": "ga:1234567890",
      "metrics": "ga:totalEvents,ga:uniqueEvents",
      "start-date": "2020-11-01",
    },
    "ga-chart-overview",
    "LINE",
    Object {
      "hAxis": Object {
        "gridlines": Object {
          "units": Object {
            "days": Object {
              "format": Array [
                "dd MMM",
              ],
            },
          },
        },
        "minorGridlines": Object {
          "units": Object {
            "hours": Object {
              "format": Array [
                "",
              ],
            },
            "minutes": Object {
              "format": Array [
                "",
              ],
            },
          },
        },
      },
    },
  ],
  Array [
    Object {
      "dimensions": "ga:date",
      "end-date": "2020-11-02",
      "filters": "ga:dimension2==editionId",
      "ids": "ga:1234567890",
      "metrics": "ga:totalEvents,ga:uniqueEvents",
      "start-date": "2020-11-01",
    },
    "ga-chart-overview",
    "LINE",
    Object {
      "hAxis": Object {
        "gridlines": Object {
          "units": Object {
            "days": Object {
              "format": Array [
                "dd MMM",
              ],
            },
          },
        },
        "minorGridlines": Object {
          "units": Object {
            "hours": Object {
              "format": Array [
                "",
              ],
            },
            "minutes": Object {
              "format": Array [
                "",
              ],
            },
          },
        },
      },
    },
  ],
]
`;
